src/attribute.cpp

```diff
@@ -27,6 +43,166 @@ Integer::Integer(uint8_t type, const uint8_t* data, size_t size)
             + data[3];
 }
 
+Integer::Integer(uint8_t type, uint32_t value)
+    : Attribute(type),
+      m_value(value)
+{
+}
+
+std::vector<uint8_t> Integer::toVector(const std::string& secret, const std::array<uint8_t, 16>& auth) const
+{
+    std::vector<uint8_t> attribute(6);
+    attribute[0] = type();
+    attribute[1] = 6;
+    attribute[2] = m_value / (1 << 24);
+    attribute[3] = (m_value / (1 << 16)) % 256;
+    attribute[4] = (m_value / (1 << 8)) % 256;
+    attribute[5] = m_value % 256;
+    return attribute;
+}
+
+IpAddress::IpAddress(uint8_t type, const uint8_t* data, size_t size)
+        : Attribute(type)
+{
+    if (size != 4)
+        throw std::runtime_error("Invalid integer attribute size. Should be 4, actual size is " + std::to_string(size));
+
+    for (size_t i = 0; i < size; ++i)
+        m_value[i] = data[i];
+}
+
+IpAddress::IpAddress(uint8_t type, const std::array<uint8_t, 4>& address)
+    : Attribute(type),
+      m_value(address)
+{
+}
+
+std::vector<uint8_t> IpAddress::toVector(const std::string& secret, const std::array<uint8_t, 16>& auth) const
+{
+    std::vector<uint8_t> attribute(6);
+
+    attribute[0] = type();
+    attribute[1] = attribute.size();
+    for (size_t i = 0; i < m_value.size(); ++i)
+        attribute[i + 2] = m_value[i];
+
+    return attribute;
+}
+
+Encrypted::Encrypted(uint8_t type, const uint8_t* data, size_t size, const std::string& secret, const std::array<uint8_t, 16>& auth)
+        : Attribute(type)
+{
+    if (size > 128)
+        throw std::runtime_error("Invalid encrypted attribute size. Should be max 128 bytes, actual size is " + std::to_string(size));
+
+    std::vector<uint8_t> mdBuffer(16 + secret.length());
+
+    for (size_t i = 0; i < secret.length(); ++i)
+        mdBuffer[i] = secret[i];
+
+    for (size_t i = 0; i < 16; ++i)
+        mdBuffer[i + secret.length()] = auth[i];
+
+    std::vector<uint8_t> plaintext(size);
+
+    for (size_t i = 0; i < size / 16; ++i)
+    {
+        std::array<uint8_t, 16> md;
+
+        MD5(mdBuffer.data(), mdBuffer.size(), md.data());
+        for (size_t j = 0; j < 16; ++j)
+            plaintext[i * 16 + j] = data[i * 16 + j] ^ md[j];
+
+        for (size_t j = 0; j < secret.length(); ++j)
```

Comment:
I think you can skip filling secret here. It is still in th buffer, MD5 does not change the source.